rpc_util.go

```diff
@@ -782,8 +782,20 @@ func decompress(compressor encoding.Compressor, d []byte, maxReceiveMessageSize
 			// size is used as an estimate to size the buffer, but we
 			// will read more data if available.
 			// +MinRead so ReadFrom will not reallocate if size is correct.
-			buf := bytes.NewBuffer(make([]byte, 0, size+bytes.MinRead))
-			bytesRead, err := buf.ReadFrom(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))
+			bufferSize := uint64(size) + bytes.MinRead
+			if bufferSize > math.MaxInt {
+				bufferSize = math.MaxInt
+			}
+			buf := bytes.NewBuffer(make([]byte, 0, int(bufferSize)))
+
+			bytesRead, err := buf.ReadFrom(io.LimitReader(dcReader, int64(maxReceiveMessageSize)))
+			if bytesRead == int64(maxReceiveMessageSize) {
+				b := make([]byte, 1)
+				if n, err := dcReader.Read(b); n > 0 || err != io.EOF {
+					return nil, size + n, fmt.Errorf("overflow: message larger than max size receivable by client (%v bytes)", maxReceiveMessageSize)
+				}
+			}
+
 			return buf.Bytes(), int(bytesRead), err
 		}
 	}
```

Comment:
We will also need fixes below here in case `DecompressedSize` is not implemented by the compressor.

As @arvindbr8 mentioned, we'll also want tests for overflowing the max message size with both kinds of compressors, too (some may already exist), and also covering the case where the max size is set to `MaxInt`.  I guess we can't test overflowing with `MaxInt` very easily, though.  Maybe 2GB (on a 32-bit system) isn't asking _too_ much(?), but even that could be problematic.

I think a unit test of `decompress` would be ideal here, with an internal way to override `MaxInt` to a smaller value.

Reply:
hi @dfawley,
regarding the fixes for `compressor` not implementing `DecompressedSize`, could you elaborate on what the intended behaviour should be?
I had the following changes in mind:
```
if sizer, ok := compressor.(interface {
	DecompressedSize(compressedBytes []byte) int
}); ok {
	...
} else {
	log.Printf("warn: compressor does not implement method DecompressedSize()")
}
// Read from LimitReader with limit max+1. So if the underlying
// reader is over limit, the result will be bigger than max.
d, err = io.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))

if len(d) > maxReceiveMessageSize {
	return nil, len(d), nil  // same response signature as when DecompressedSize() is implemented
}
return d, len(d), err

```

Reply:
Sure, something like that LGTM.  It would be great if you can refactor the code to remove the duplication, though, since both code paths are very similar.